### PR TITLE
#10026 Implement DocumentSymbolProviderMetadata

### LINKED ISSUE #10026: Implement `DocumentSymbolProviderMetadata`
### Feature Description:
Concretely, there is now a label for the provider. When registering a provider, we should pass this information all the way to the monaco editor. We need to use that label in `monaco-outline-contribution.ts`

See https://code.visualstudio.com/api/references/vscode-api#DocumentSymbolProviderMetadata

### PR DESCRIPTION
**Pull request by ndoschek**
#### What it does

This PR adds the implementation for `DocumentSymbolProviderMetadata` and uses label in `MonacoOutlineContribution`:
- Introduce DocumentSymbolProviderMetadata interface in theia.d.ts
- Adapt register functions to pass on optional metadata
- Adapt the MonacoOutlineContribution and use metadata label or plugin's displayName as root node if multiple providers are registered (aligned to behavior in VSCode)
- Remove Theia's custom DocumentSymbolProvider for 'launch.json' files and reuse VSCode's provider (available via vscode-builtin-configuration-editing)

Contributed on behalf of STMicroelectronics

Signed-off-by: Nina Doschek <ndoschek@eclipsesource.com>

Fixes #10026

#### How to test

1. Open a `launch.json` file in your workspace (e.g. https://github.com/eclipse-theia/theia/blob/master/.vscode/launch.json)
2. Open the Outline View via menu `View` > `Outline`
3. Comparable to VSCode, there are two providers registered:
   1. The plugin `vscode-builtin-json-language-features`
       Provides the JSON document outline under the root node `JSON Language Features (built-in)`
   2. The plugin `vscode-builtin-configuration-editing`
       Provides all launch configs as document symbols under the root node `Launch Targets`.
4. It is shown that either [the metadata label (in case of 3.ii.)](https://github.com/microsoft/vscode/blob/0a644fc76517b1aeeb37b8fdf8d636d39a3d9474/extensions/configuration-editing/src/configurationEditingMain.ts#L109) is used as root node display name or if not available the plugin display name (in case of 3.i.).
      

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by msujew** [comment]
@colin-grant-work Can you tell, whether this change is in conflict with your changes for the monaco uplift? If so, we should probably wait for your uplift to finish before merging/reviewing other monaco related changes.

**Comment by colin-grant-work** [comment]
@msujew, at a glance, it looks like it'll only require modifying the references to `monaco.Range` to use imports to be compatible with the uplift, so I don't think it needs to wait - fixing the uplift once this is merged will be straightforward.

File: packages/plugin-ext/src/plugin/languages.ts [hunk 1]
```diff
@@ -531,12 +531,18 @@ export class LanguagesExtImpl implements LanguagesExt {
...
         const callId = this.addNewAdapter(new OutlineAdapter(this.documents, provider));
-        this.proxy.$registerOutlineSupport(callId, pluginInfo, this.transformDocumentSelector(selector));
+        const displayName = (metadata && metadata.label) || LanguagesExtImpl._pluginLabel(pluginInfo);
+        this.proxy.$registerOutlineSupport(callId, pluginInfo, this.transformDocumentSelector(selector), displayName);
         return this.createDisposable(callId);
     }
 
+    private static _pluginLabel(pluginInfo: PluginInfo): string {
+        return pluginInfo.displayName || pluginInfo.name;
+    }
```
Comment (msujew) [inline packages/plugin-ext/src/plugin/languages.ts:544]: Suggestion: Having this as a `private static` method feels a bit out of place. How do you think about moving it to the other utility functions to the bottom of the file?

**Review by msujew — COMMENTED** [review]
The changes are looking really good. Happy that we're able to remove the adhoc document symbol provider for the `launch.json` :)

- [x] Opening the `launch.json` file shows two distinct document symbol providers
- [x] Other `*.json` files show the usual json symbols
- [x] Symbols for code (tested with `*.ts`) work as expected

A minor comment before this is ready to be merged.

File: packages/plugin-ext/src/plugin/languages.ts (hunk 1 above)
Comment (ndoschek) [reply packages/plugin-ext/src/plugin/languages.ts:544]: Yes, definitely! Thanks for the quick review!

**Review by msujew — APPROVED** [review]
Thanks, looking good!

**Pull request merged**